### PR TITLE
NeuronORPOTrainer

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

Adds initial support for the [ORPOTRainer](https://huggingface.co/docs/trl/v0.11.4/en/orpo_trainer) but without parallelism support.

### CODE REVIEW

**Comment by dacorvo** [comment]
@michaelbenayoun the CI is failing with your changes: can you check what went wrong so that we can start reviewing the pull-request ?

File: optimum/neuron/trainers.py [hunk 1]
```diff
@@ -1863,3 +1874,331 @@ def tokenize(element):
...
+        self,
+        policy_chosen_logps: torch.FloatTensor,
+        policy_rejected_logps: torch.FloatTensor,
+    ) -> Tuple[torch.FloatTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
+        # Derived from Eqs. (4) and (7) from https://huggingface.co/papers/2403.07691 by using log identities and exp(log(P(y|x)) = P(y|x)
+        # log_odds = (policy_chosen_logps - policy_rejected_logps) - (
+        #     torch.log1p(-torch.exp(policy_chosen_logps)) - torch.log1p(-torch.exp(policy_rejected_logps))
+        # )
+
+        # We used this instead of `torch.log1p` because it produces NaNs in BF16.
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:2189]: Can you make it clearer it is something neuron specific ?

**Review by dacorvo — COMMENTED** [review]
Thanks for the pull-request ! LGTM, but I would like the changes to the original trl code to appear more clearly.

**Comment by michaelbenayoun** [comment]
Can I merge?

**Review by dacorvo — APPROVED** [review]
LGTM, thanks !

**Pull request merged**